### PR TITLE
[Merged by Bors] - feat: Port Data.Set.Semiring

### PR DESCRIPTION
**Pull request by riccardobrasca**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Semiring.lean [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  ⟨fun _ _ _ => mul_subset_mul_right⟩
+#align set_semiring.covariant_class_mul_right SetSemiring.covariant_class_mul_right
+
+end Mul
+
+-- Porting note: this was `one := 1`
+instance [MulOneClass α] : NonAssocSemiring (SetSemiring α) :=
+  { (inferInstance : NonUnitalNonAssocSemiring (SetSemiring α)),
+    Set.mulOneClass with
+    one := Set.up ({1} : Set α)
```
Comment (riccardobrasca) [inline Mathlib/Data/Set/Semiring.lean:142]: Originally this was `one := 1`, but it doesn't work with
```bash
failed to synthesize instance
  OfNat (SetSemiring α) 1
```

File: Mathlib/Data/Set/Semiring.lean [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+
+open Function Set
+
+open Pointwise
+
+variable {α β : Type _}
+
+-- Porting note: mathlib3 uses `deriving Inhabited, PartialOrder, OrderBot`
+/-- An alias for `set α`, which has a semiring structure given by `∪` as "addition" and pointwise
```
Comment (ChrisHughes24) [inline Mathlib/Data/Set/Semiring.lean:29]:
```suggestion
/-- An alias for `Set α`, which has a semiring structure given by `∪` as "addition" and pointwise
```

File: Mathlib/Data/Set/Semiring.lean [hunk 3]
```diff
@@ -25,120 +25,143 @@ open Pointwise
 
 variable {α β : Type _}
 
+-- Porting note: mathlib3 uses `deriving Inhabited, PartialOrder, OrderBot`
 /-- An alias for `set α`, which has a semiring structure given by `∪` as "addition" and pointwise
```
Comment (jcommelin) [inline Mathlib/Data/Set/Semiring.lean]:
```suggestion
/-- An alias for `Set α`, which has a semiring structure given by `∪` as "addition" and pointwise
```

**Review by jcommelin — COMMENTED** [review]
The fact that dot notation no longer works is very mysterious.

File: Mathlib/Data/Set/Semiring.lean (hunk 3 above)
Comment (riccardobrasca) [reply Mathlib/Data/Set/Semiring.lean]: This is already done.. maybe you were having a look to an old commit.

**Comment by riccardobrasca** [comment]
bors merge

**Pull request closed without merging**